### ISSUE TITLE
[web] Precise scrolling impossible by default

### ISSUE THREAD

**Issue by creativecreatorormaybenot** [flutter/flutter#83368]
This is mostly a usability concern / feature request rather than a bug report. This is because it is possible to work around it (see end of comment), but it is honestly a bad experience.

## Problem

Based on the scrolling settings for the mouse / device used, scrolling using the scroll wheel might be very extreme (how many lines to scroll each time). This means that on my device, it is **impossible** for me to view some of the content in a Flutter app.

Here is an example screen recording of what that looks like (my setting is `58` lines scrolled at a time, max would be `100`). I am scrolling the **minimum amount** each time (hitting the scroll wheel once):

https://user-images.githubusercontent.com/19204050/119557020-36a9c200-bd8f-11eb-8ffa-64923e1a9398.mp4

As you can see, it is impossible for me to use the Flutter Gallery web app in this case (without exaggerating - on this device, there is no way for me to do it without changing hardware or system settings). See https://gallery.flutter.dev/#/ for the app running on web.

## Solution

As the web is a mature platform, there are of course many pre-existing solutions to this problem.

### Middle mouse button scrolling

In this video, you can see that the same problem exists on normal HTML web apps (https://lipsum.com) at the start. However, I can then press the middle mouse button to bring up a precision tool for smooth scrolling.

https://user-images.githubusercontent.com/19204050/119557578-e41cd580-bd8f-11eb-85d9-6bfe2870b959.mp4

In Flutter web, this is simply a noop (pressing the middle mouse button does not bring up anything by default). You can see a separate feature request for this: https://github.com/flutter/flutter/issues/66537

### Default scrollbar

In this recording, you can see that regular web pages have a scrollbar that allows precise scrolling **by default**.

https://user-images.githubusercontent.com/19204050/119557929-54c3f200-bd90-11eb-91d7-40148b92d82b.mp4

In Flutter, we _also_ have default scrollbars since #78588. However, there do not allow dragging them to scroll at all.

### Text selection (edge scrolling)

This one seems a bit more "hackish", but it is definitely a core feature in web browser scrolling and I do often find myself using it. You can see that on the normal HTML page, I can select the at the bottom of the page to precisely scroll to the next line at the bottom of the screen:

https://user-images.githubusercontent.com/19204050/119558407-e03d8300-bd90-11eb-805e-020f01536c7a.mp4

In Flutter web, this does not work at all. `SelectableText` is barely functional on web in this way anyway (you cannot scroll across multiple widgets) and definitely not beyond the edge of the screen.  
I have found an old issue by @LongCatIsLooong (#32344) for **edge scrolling** on mobile (for text fields), however, I have not found an issue for edge scrolling on desktop. Should we create that?

## Workaround

Before #81569, Flutter web apps used to allow scrolling using drag by default (which is why I was able to use Flutter web apps on this device). Now, this is not possible _as a workaround_ anymore (by default).

You can still manually apply this workaround using:

```dart
ScrollConfiguration(
  behavior: ScrollConfiguration.of(context).copyWith(
    dragDevices: {
      PointerDeviceKind.mouse,
      PointerDeviceKind.touch,
    },
  ),
  child: ScrollView(...),
)
```

However, this is only a subpar workaround and does not address the issue that precise scrolling is impossible by default, i.e. Flutter apps created by developers who do not pay attention to this detail, I will not be able to use on this device (like Flutter Gallery).  
I have created this issue to collect all of this - of course it would be best if all 3 solutions that normal web apps support would work in Flutter as well, but solving this at all would be a start.

**Comment by jonahwilliams** [comment]
See also: https://github.com/flutter/flutter/issues/32120

**Comment by creativecreatorormaybenot** [comment]
Thanks @jonahwilliams - I assume this explains why it seems more eradic than on the rest of the web 👍 But I would assume the usability stays the same because you cannot *end* at any of the steps in between, right?

**Comment by jonahwilliams** [comment]
The large jump in your first example feels like a case where there is some transform applied to the screen or scrollable that is not applied to the scroll delta, since the jump is so large.

In general a mouse wheel does not give perfectly precise scrolling unless you toggle the drag scroll mode which I don't think we support as a gesture.

**Comment by creativecreatorormaybenot** [comment]
>The large jump in your first example feels like a case where there is some transform applied to the screen or scrollable that is not applied to the scroll delta, since the jump is so large.

I think it is just because of my scroll settings. On Windows, e.g. it is called "Change how far you scroll with the mouse wheel" and you can select a value between 1 to 100 lines. The default is `3`, which is what Flutter is likely tested against.

**Comment by adrianflutur** [comment]
Hi @creativecreatorormaybenot, I also experienced theese scrolling limitations, especially the middle mouse scrolling one.

So, I tried to implement every missing scrolling feature I needed for Web and Desktop in this package:
https://github.com/adrianflutur/flutter_improved_scrolling

After a lot of tweaking I actually find it pretty good at simulating the "native" scrolling experience one would get on Web or on Desktop. 
Maybe you should give it a shot until the whole scrolling mechanism is fixed inside the framework itself 😀

**Comment by Piinks** [comment]
Hey @creativecreatorormaybenot can you please split this into multiple issues or just follow the already existing issues you have referenced? There is guidance about [making issues actionable](https://github.com/flutter/flutter/wiki/Issue-hygiene#every-issue-should-be-actionable) in the [Issue Hygiene](https://github.com/flutter/flutter/wiki/Issue-hygiene) page of the wiki.

I am going to try to address all of the various issues you have listed here. They should each have an individual issue though with reproducible information so that they can be addressed.

- > On Windows, e.g. it is called "Change how far you scroll with the mouse wheel" and you can select a value between 1 to 100 lines. The default is 3, which is what Flutter is likely tested against.
  - If we do not have an issue for this specific piece, please file one with a minimal reproduction and the output of `flutter doctor -v`
- Middle mouse button scrolling
  - As mentioned, an issue already exists for this feature, please follow #66537 for updates.
- Default scrollbar dragging
  - Scrollbars already support dragging. If there is a use case where this functionality is broken, please file an issue with a minimal reproduction and the output of `flutter doctor -v`. Please do tag me and I will take a look.
- Text selection (edge scrolling)
  - The should definitely be supported on desktop and web. I have reopened #32344 and updated it to reflect these platform requirements. Please follow that issue for updates.

Once everything is accounted for, this issue should be closed in lieu of following the associated issues. This issue in its current state is not specific or actionable, especially for those parts that really need a reproduction in order to identify a fix.

At times, we create umbrella (☂️ ) issues like https://github.com/flutter/flutter/issues/90559 as planned work. These represent a collection individual issues that all have the necessary information to be actionable. These issues are typically created for a related area of work, and serve as a quick reference point for tracking progress. They are always assigned when created, because someone is already assigned to the collection of issues that are being addressed. I do not recommend creating umbrella issues.

Feel free to update this when each issue has its own tracker. Thanks!

**Comment by creativecreatorormaybenot** [comment]
@Piinks Thanks for getting back here.

This issue was created in response to other, specific issues not being properly accounted for. For example:  
Since the change of the default scrolling behavior on desktop, we cannot use Flutter web apps by default anymore. Instead, we have to always specify:

```dart
scrollBehavior: ScrollConfiguration.of(context).copyWith(
    dragDevices: PointerDeviceKind.values.toSet(),
  ),
```

In our web apps.

This is because of https://github.com/flutter/flutter/issues/83359, which you closed unfortunately. There has been no improvement so far in terms of that issue and I would argue it is specific.  
There is one problem about it: you and me, we probably do not have access to the devices that do not work with the new scrolling at all (because the trackpads on these devices work differently).  
However, I still cannot scroll horizontally in Flutter web apps on my desktop Windows computer by default (though that has an issue https://github.com/flutter/flutter/issues/75180).

That was basically the _trigger_ for this issue.

---

Regarding:

> On Windows, e.g. it is called "Change how far you scroll with the mouse wheel" and you can select a value between 1 to 100 lines. The default is 3, which is what Flutter is likely tested against.

That is not an issue - this works as expected (but that is the problem). On my computer, I cannot really scroll vertically in Flutter web apps as well because it will skip half of the page.  
In regular web apps, I use the **scroll bar for dragging** or I use the middle mouse button for scrolling (https://github.com/flutter/flutter/issues/66537).

In our web app, you **cannot drag scroll bars** unless specifying:

```dart
scrollBehavior: ScrollConfiguration.of(context).copyWith(
    dragDevices: PointerDeviceKind.values.toSet(),
  ),
```

I am not sure if you are suggesting it works by default. In that case, it does not work.

---

Regarding _text selection (edge scrolling)_ - thanks for reopening https://github.com/flutter/flutter/issues/32344.

However, I guess that is a bit misleading because the default HTML behavior is **not** about **text fields**. Instead, it is about the **text you read**, i.e. the default text on any web page.

Also, I would have two action items here:

1. Could you _unlock the discussion_ on https://github.com/flutter/flutter/issues/32344 🙏?
2. Should I create a new issue for **edge scrolling** _outside of text fields_? :v:

---

@Piinks Thanks a lot for the effort and considering this. I still feel like the three topics (between the three dividers above) are not properly addressed with the current issues. Let me know how you would follow up on each of these topics.

And most importantly, this issue is about the **default** behavior. I feel like it is a major problem that a Flutter web app behaves completely unlike a regular web app **by default**. I think you can make it feel a lot more like a regular web app if you deliberately put effort into it, but we should make this the default.

**Comment by Piinks** [comment]
> This issue was created in response to other, specific issues not being properly accounted for. 

This is specifically discouraged [in our wiki](https://github.com/flutter/flutter/wiki/Issue-hygiene#exceptions):

> **_Intentional duplicates like "This is the same as bug #ABC but that one is not getting enough attention." Instead, upvote the original issue or add a comment that provides new details that are not already captured or (best of all) assign it to yourself and start working on it!_**

---

> For example: Since the change of the default scrolling behavior on desktop, we cannot use Flutter web apps by default anymore. Instead, we have to always specify `[...]` in our web apps.
> 
> This is because of #83359, which you closed unfortunately. 

Similarly to this issue, that issue duplicated existing issues and was closed with direction to follow the specific issues that already existed. 

> There has been no improvement so far in terms of that issue

That is because it is closed. :) If you refer to the issues it references, it looks like trackpad support is currently in progress.

> However, I still cannot scroll horizontally in Flutter web apps on my desktop Windows computer by default (though that has an issue #75180).

So please follow that issue. Filing duplicate issues to try to draw more attention to existing issues detracts from any effort, conversation or collaboration that may be happening on the already existing issue.

---
 
> > On Windows, e.g. it is called "Change how far you scroll with the mouse wheel" and you can select a value between 1 to 100 lines. The default is 3, which is what Flutter is likely tested against.
> 
> That is not an issue - this works as expected (but that is the problem). 

I am not sure how to address this if you say it is working as expected? If there is work that needs to be done here, please file an issue.

---

> In our web app, you **cannot drag scroll bars** unless specifying `[...]`.

This sounds like an issue with your app. You should not need to alter the ScrollBehavior in order to drag a scrollbar, please file an issue. Disabling drag on desktop relates to dragging a Scrollable with a mouse, it does not affect the scrollbar.

---
 
> Regarding _text selection (edge scrolling)_ - thanks for reopening #32344.
> 
> However, I guess that is a bit misleading because the default HTML behavior is **not** about **text fields**. Instead, it is about the **text you read**, i.e. the default text on any web page.
> 
> Also, I would have two action items here:
> 
> 1. Could you _unlock the discussion_ on [Implement edge scrolling for text selection in text fields #32344](https://github.com/flutter/flutter/issues/32344) 🙏?

Done.

> 2. Should I create a new issue for **edge scrolling** _outside of text fields_? ✌️

I have updated the title to reflect this. Please feel free to add your comments to the issue.

--- 

>  I still feel like the three topics (between the three dividers above) are not properly addressed with the current issues.

If the issues do not fully capture the need, please follow the guidance from Issue Hygeine, as mentioned above - **_add a comment that provides new details that are not already captured_**. 

> And most importantly, this issue is about the **default** behavior. I feel like it is a major problem that a Flutter web app behaves completely unlike a regular web app **by default**.

We cannot tackle absolutely everything that we want to see improved for Flutter Web in one issue. Keeping issues specific and actionable is how Flutter will improve.

**Comment by creativecreatorormaybenot** [comment]
@Piinks thanks 🙏 

> That is because it is closed. :) [...]
> Similarly to this issue, that issue duplicated existing issues and was closed with direction to follow the specific issues that already existed.

> add a comment that provides new details that are not already captured.

Two notes about this:

1. The issues you mentioned in [your comment](https://github.com/flutter/flutter/issues/83359#issuecomment-853211557) do not address the issue I outlined (i.e. horizontal scroll not working on some devices at all).
2. I cannot comment anymore. I generally feel the auto-locking GitHub bot is not productive, but yeah - I cannot comment.

So regarding #83359, it was closed but there is no other issue for this afaic.

---

The point of this issue is not to be an "Intentional duplicate" - instead, I think the `scrollingBehavior` (with the current Flutter web feature set) should be different. You can close it if you disagree, but it is not a duplicate.

The reason I elaborated so much in the original comment was to make it understandable.

**Comment by creativecreatorormaybenot** [comment]
@Piinks Some input for you: I did not find any of the comments helpful.

Regarding the last suggestion, I do not understand what `flutter/tests` has to do with this. Scrolling does not work on some Laptops (for hardware-related reasons). It is impossible to test this with Flutter unit tests.

And even if your suggestion that this will be fixed with #23604 (which I would certainly challenge) turns out to be true - there is still no way of testing it **at least** until the feature is implemented in the framework because emulating the input device is **inherently** unsupported for the time being.
